File: 1 - Multiples of 3 and 5/multiples.ex
defmodule Multiples do

  @moduledoc"""
    If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9.
    The sum of these multiples is 23.

    Find the sum of all the multiples of 3 or 5 below 1000.
  """

  # Solution with recursion
  def sum(x, max) when x >= max do 0 end
  def sum(x, max) when rem(x,3) == 0 or rem(x,5) == 0 do
    x + sum(x + 1, max)
  end
  def sum(x, max) do
    sum(x + 1, max)
  end

  # Solution with enums
  def enum(x, max) do
    Enum.sum(
      Enum.filter(x..max-1, fn x -> rem(x,3) == 0 or rem(x,5) == 0 end)
    )
  end

end

IO.puts(Multiples.sum(1, 1000))

#IO.puts(Multiples.enum(1,1000))

File: 1 - Multiples of 3 and 5/multiples.rs
// https://projecteuler.net/problem=1
//
// If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.
//
// Find the sum of all the multiples of 3 or 5 below 1000.
//

fn multiplesi(x: i32) -> i32 {
    let mut sum: i32 = 0;
    for i in 1..x + 1 {
        if i % 3 == 0 || i % 5 == 0 {
            sum += i;
        }
    }
    sum
}

fn multiplesr(x: i32) -> i32 {
    if x == 1 {
        return 0;
    }
    (x % 3 == 0 || x % 5 == 0) as i32 * x + multiplesr(x - 1)
}

fn main() {
    println!("{}", multiplesi(9));
    println!("{}", multiplesr(9));
}
